src/components/JSONEdit.tsx

```diff
@@ -39,26 +42,39 @@ const JSONEdit: React.FC<JSONEditProps & TextareaHTMLAttributes<HTMLTextAreaElem
     let updatedData: any;
     try {
       updatedData = JSON.parse(rawValue);
-      setJSONValid(true);
-    } catch {
-      setJSONValid(false);
+      setJsonLogicError('');
+    } catch (error) {
+      setJsonLogicError(error.toString());
       return;
     }
-
-    updateValue(updatedData);
+    const error = validateLogic(updatedData);
+    if (error) {
+      setJsonLogicError(error);
+      return;
+    }
+    updateValue(updatedData); // valid JSON
```

Comment:
```suggestion
    updateValue(updatedData); // valid JSON & valid logic expression
```
because of `validateLogic` returns an error message, we don't run `updateValue` so the inverse must be true :)